### DISCUSSION TITLE
OpenSSL FIPS module distribution

### DISCUSSION THREAD

**Discussion by andriy-fedechko-imprivata** [openssl/openssl#25036] [Q&A] [answered]
Hello,

Can someone suggest, or provide a link to any documentation that can spread the light on the OpenSSL FIPS module distribution.
That's clear from the document that in order to be FIPS compliant there need to have config file generated and in order to generate it the fipsinstall command should be used.

That appears not that handy when you need to distribute your product/application among multiple endpoints. Basically, that will require openssl.exe be added into e.g. installation package (that might have some licensing restrictions), modifying an installation script so that it can run openssl.exe /fipsinstall command, etc.

If the above is the only option - that's also fine, not that handy but can be handled. However, that's not very clear if that's the case.

There is a comment left here: https://mta.openssl.org/pipermail/openssl-users/2022-February/014887.html that says that fips module and it's config file can be distributed across different machines but it will required that config file doesn't have configuration checksum provided, therefore that self test will be always executed upon fips module loading.

From the other hand, there is another comment: https://github.com/openssl/openssl/issues/20559#issuecomment-1479389290 that tells different thing.

So in essence - what is the correct way of distributing fips module and are there any documentation that can be followed as a reference/source of truth?

**Comment by kulkarniamit** [comment, marked as answer]
I believe the ultimate source of truth is [FIPS security policy document (3.0.9)](https://openssl-library.org/source/fips-doc/openssl-3.0.9-security-policy-2024-01-12.pdf).

To be compliant, we need to follow the exact instructions in Appendix A of the security policy:

* "_To install the FIPS configuration file to a non-default location, this can be achieved by running the
‘fipsinstall’ command line application manually_": `$ openssl fipsinstall`

* "_Note: The Module shall have the self-tests run, and the **Module config file output generated** on each platform where it is intended to be used. The Module config file output data **shall not be copied** from one machine to another_"

If we follow https://mta.openssl.org/pipermail/openssl-users/2022-February/014887.html and copy the fipsmodule.cnf without the install checksum, FIPS module will perform the self tests on each start up. This satisfies the operational requirement, but it wouldn't be compliant with the security policy letter. The letter makes it clear that the module config file output needs to be generated on each platform where it is intended to be used.
So, `fipsinstall` needs to be run on every installation before you can use the FIPS provider in a compliant way.

Consulting a FIPS validation lab might be another option for authoritative answers.

**Reply by paulidale** [reply]
Spot on.  To adhere to the letter of the standard, you must run fipsinstall on each machine (you don't need to retain the openssl.exe after this).  To adhere to the spirit of the standard, you can leave out the install checksum and the self tests run on each load.

This will be changing in the 140-3 validation to self test on every load.

**Reply by andriy-fedechko-imprivata** [reply]
Thank you for the replies. 

So in other words - if follow instructions described here: https://mta.openssl.org/pipermail/openssl-users/2022-February/014887.html my application will not be FIPS compliant? Is that right statement? 

Also, @paulidale, can you please elaborate a bit more regarding "This will be changing in the 140-3 validation to self test on every load." What exactly is going to change?

**Reply by paulidale** [reply]
Correct, your application would not be strictly compliant because you aren't following the install rules specified in appendix A of the security policy.

For 140-3  the self tests will always run on every module load.  The installation instructions will be updated to account for this and the don't copy requirement is likely going to be lifted.  I'm not privy to the submitted security policy, so don't know the precise change.  You'll have to wait for the validation to complete.

**Comment by slontis** [comment]
In general the FIPS self tests need to run before using the FIPS module.
This normally  means that self tests run every time the module is loaded.
The FIPS 140-2 IG rules allowed you to run the self tests ONCE on install, and this was done by fipinstall running the self tests and then writing this status to the config file. So you cant just run the self tests on one machine and then copy the config because it would not run the self tests on that machine.
So if you are happy to have the self tests run on each machine on every load then you could omit the fields from the config file and then copying the config would be fine.

In FIPS 140-3 the rule allowing you to run once has been taken out. So this issue with running fipsinstall on each machine will not be required when we have a FIPS 140-3 validation.

**Reply by andriy-fedechko-imprivata** [reply]
Could this also have an effect on how installation process is going to happen? E.g. getting rid of the need of running _openssl fipsinstall_ on each machine.  

Also, I'm bit confused with this statement: 
_So you cant just run the self tests on one machine and then copy the config because it would not run the self tests on that machine._
According to description here: https://mta.openssl.org/pipermail/openssl-users/2022-February/014887.html that's is what is happening if there are no config file available, in other words, if install command was not executed on target machine.

**Reply by paulidale** [reply]
For the existing 140-2 validated FIPS module, you *must* follow the instructions exactly as per the security document to be compliant.  This means running fipsinstall on each machine.  If you are willing to accept compliance within the intent, but not the letter, of the rules, things are different.

The 140-3 validation will be different.  But it's still a long way off in all likelihood.

**Reply by slontis** [reply]
It would also be possible to update the security policy (which requires lab work), so that you could also do it by removing the config items, or allow you to specify the fipsinstall option, (the security policy points you towards these options for some reason).
The original intention was to allow either.
In future validations (if it is possible??), I would suggest pointing to a link containing instructions, rather than embedding it in the security policy (so it could be updated easier).
